subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -395,48 +658,74 @@ _set_au_boundaries (GstRtpH266Pay * rtph266pay)
 static GstFlowReturn
 _push_pending_data (GstRtpH266Pay * rtph266pay, gboolean eos)
 {
+  AggregatedNalus *aggregated_nalus = &rtph266pay->aggregated_nalus;
   GQueue *nalus = &rtph266pay->nalus;
   GstFlowReturn ret = GST_FLOW_OK;
   Nalu *nalu;
 
-  // TODO: handle AP
-  // TODO: send XPS if needed
-
-  // TODO: Maybe it's worth to avoid iterating the NALU list twice, maybe not
+  // FIXME: Maybe it's worth to avoid iterating the NALUs several times, maybe not
+  _insert_ps_cache (rtph266pay);
   _set_au_boundaries (rtph266pay);
 
-  // Try to push all NALUs
+  // 1. Aggregate NALUs while possible
+  // ---- vvv Can't aggregate anymore vvv ----
+  // 2. If it has been possible to **aggregate something**:
+  //   1. Keep the current NALU, it'll be processed later
+  //   2. Send the current Aggregation Packet
+  // 3. If it has been possible to **aggregate nothing**:
+  //   * If the packet fits into a MTP -> Send a Unit Packet
+  //   * Else -> Send two or more Fragmentation Units
+  //
+  // Note: _push_aggregated will send a UP when there is only 1 NALU aggregated
   while ((nalu = g_queue_pop_head (nalus))) {
-    gboolean is_last = g_queue_get_length (nalus) == 0;
-
-    // Can't push the last NALU without knowing if it's the end of an AU,
-    // because setting the M bit could be necessary. But have to push it if
-    // we're on EOS
-    if (!eos && is_last && !nalu->au_end) {
-      GST_DEBUG_OBJECT (rtph266pay, "Keeping last NALU: %" NALU_PTR_FORMAT,
+    // Aggregate NALUs while possible
+    if (_can_aggregate_nalu (rtph266pay, nalu)) {
+      _aggregated_nalus_add (aggregated_nalus, nalu);
+      GST_DEBUG_OBJECT (rtph266pay, "NALU aggregated: %" NALU_PTR_FORMAT,
           NALU_ARGS (nalu));
-      g_queue_push_head (nalus, nalu);
-      break;
+      continue;
     }
 
-    GST_DEBUG_OBJECT (rtph266pay, "Pushing NALU: %" NALU_PTR_FORMAT,
-        NALU_ARGS (nalu));
+    // vvv Can't aggregate anymore vvv
+    if (!_aggregated_nalus_is_empty (aggregated_nalus)) {
+      g_queue_push_head (nalus, nalu);  // Keep this one for later
+      ret = _push_aggregated (rtph266pay);      // Push what we have right now
+      if (ret != GST_FLOW_OK)
```

Comment:
Handle the case when ret is not GST_FLOW_OK.

Reply:
:confused: :question:  :man_shrugging: